File: modules/backlogs/app/models/impediment.rb
#-- copyright
# OpenProject is an open source project management software.
# Copyright (C) the OpenProject GmbH
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License version 3.
#
# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
# Copyright (C) 2006-2013 Jean-Philippe Lang
# Copyright (C) 2010-2013 the ChiliProject Team
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
# See COPYRIGHT and LICENSE files for more details.
#++

class Impediment < Task
  extend OpenProject::Backlogs::Mixins::PreventIssueSti

  before_save :update_blocks_list

  validate :validate_blocks_list

  def self.default_scope
    roots
      .where(type_id: type)
  end

  def blocks_ids=(ids)
    @blocks_ids = [ids] if ids.is_a?(Integer)
    @blocks_ids = ids.split(/\D+/).map(&:to_i) if ids.is_a?(String)
    @blocks_ids = ids.map(&:to_i) if ids.is_a?(Array)
  end

  def blocks_ids
    @blocks_ids ||= blocks_relations.map(&:to_id)
  end

  private

  def update_blocks_list
    mark_blocks_to_destroy

    build_new_blocks
  end

  def validate_blocks_list
    if blocks_ids.empty?
      errors.add :blocks_ids, :must_block_at_least_one_work_package
    else
      other_version_ids = WorkPackage.where(id: blocks_ids).pluck(:version_id).uniq
      if other_version_ids.size != 1 || other_version_ids[0] != version_id
        errors.add :blocks_ids,
                   :can_only_contain_work_packages_of_current_sprint
      end
    end
  end

  def mark_blocks_to_destroy
    blocks_relations.reject { |relation| blocks_ids.include?(relation.to_id) }.each(&:mark_for_destruction)
  end

  def build_new_blocks
    (blocks_ids - blocks_relations.select { |relation| blocks_ids.include?(relation.to_id) }.map(&:to_id)).each do |id|
      blocks_relations.build(to_id: id)
    end
  end
end

File: frontend/src/app/modules/augmenting/dynamic-scripts/backlogs/impediment.js
//-- copyright
// OpenProject is an open source project management software.
// Copyright (C) 2012-2020 the OpenProject GmbH
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License version 3.
//
// OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
// Copyright (C) 2006-2017 Jean-Philippe Lang
// Copyright (C) 2010-2013 the ChiliProject Team
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// See docs/COPYRIGHT.rdoc for more details.
//++

/**************************************
  IMPEDIMENT
***************************************/

RB.Impediment = (function ($) {
  return RB.Object.create(RB.Task, {

    initialize: function (el) {
      var j;  // This ensures that we use a local 'j' variable, not a global one.

      this.$ = j = $(el);
      this.el = el;

      j.addClass("impediment"); // If node is based on #task_template, it doesn't have the impediment class yet

      // Associate this object with the element for later retrieval
      j.data('this', this);

      j.on('mouseup', '.editable', this.handleClick);
    },

    // Override saveDirectives of RB.Task
    saveDirectives: function () {
      var j, prev, statusID, data, url;

      j = this.$;
      prev = this.$.prev();
      statusID = j.parent('td').first().attr('id').split("_")[1];

      data = j.find('.editor').serialize() +
                 "&is_impediment=true" +
                 "&version_id=" + RB.constants.sprint_id +
                 "&status_id=" + statusID +
                 "&prev=" + (prev.length === 1 ? prev.data('this').getID() : '') +
                 (this.isNew() ? "" : "&id=" + j.children('.id').text());

      if (this.isNew()) {
        url = RB.urlFor('create_impediment', {sprint_id: RB.constants.sprint_id});
      }
      else {
        url = RB.urlFor('update_impediment', {id: this.getID(), sprint_id: RB.constants.sprint_id});
        data += "&_method=put";
      }

      return {
        url: url,
        data: data
      };
    }
  });
}(jQuery));
